File: app/src/main/java/ru/p3tr0vich/fuel/YandexMapJavascriptInterface.java
package ru.p3tr0vich.fuel;

import android.app.Activity;
import android.os.Build;
import android.support.annotation.NonNull;
import android.webkit.JavascriptInterface;
import android.webkit.WebView;

@SuppressWarnings("unused")
class YandexMapJavascriptInterface {

    public static final String NAME = "YandexMapJavascriptInterface";

    private static final String JS = "javascript:";

    private final Activity mActivity;
    private final WebView mWebView;

    public interface YandexMap {

        double getMapCenterLatitude();

        double getMapCenterLongitude();

        String getStartSearchControlPlaceholderContent();

        String getFinishSearchControlPlaceholderContent();

        String getEmptyBalloonContent();

        int getStartSearchControlLeft();

        int getStartSearchControlTop();

        int getFinishSearchControlLeft();

        int getFinishSearchControlTop();

        void onDistanceChange(int distance);

        void onMapCenterChange(String text, String title, String subtitle,
                               double latitude, double longitude);

        void onEndLoading(boolean hasError);

        void onGeolocationStart();

        void onGeolocationFinish();

        void onErrorSearchPoint();

        void onErrorConstructRoute();

        void onErrorGeolocation();
    }

    YandexMapJavascriptInterface(@NonNull Activity activity, @NonNull WebView webView) {
        if (!(activity instanceof YandexMap))
            throw new ImplementException(activity, YandexMap.class);

        mActivity = activity;
        mWebView = webView;
    }

    @NonNull
    private YandexMap getYandexMapActivity() {
        return (YandexMap) mActivity;
    }

    private void runJavaScript(@NonNull String script) {
        script = JS + script;

//        UtilsLog.d(NAME, "runJavaScript", "script == " + script);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT)
            mWebView.evaluateJavascript(script, null);
        else
            mWebView.loadUrl(script);
    }

    public void setStartLocation(double latitude, double longitude) {
        runJavaScript("setStartLocation(" + String.valueOf(latitude) + ", " + String.valueOf(longitude) + ")");
    }

    public void setZoom(boolean inc) {
        runJavaScript("setZoom(" + inc + ")");
    }

    @JavascriptInterface
    public double getMapCenterLatitude() {
        return getYandexMapActivity().getMapCenterLatitude();
    }

    @JavascriptInterface
    public double getMapCenterLongitude() {
        return getYandexMapActivity().getMapCenterLongitude();
    }

    @JavascriptInterface
    public int getStartSearchControlLeft() {
        return getYandexMapActivity().getStartSearchControlLeft();
    }

    @JavascriptInterface
    public int getStartSearchControlTop() {
        return getYandexMapActivity().getStartSearchControlTop();
    }

    @JavascriptInterface
    public int getFinishSearchControlLeft() {
        return getYandexMapActivity().getFinishSearchControlLeft();
    }

    @JavascriptInterface
    public int getFinishSearchControlTop() {
        return getYandexMapActivity().getFinishSearchControlTop();
    }

    @JavascriptInterface
    public String getStartSearchControlPlaceholderContent() {
        return getYandexMapActivity().getStartSearchControlPlaceholderContent();
    }

    @JavascriptInterface
    public String getFinishSearchControlPlaceholderContent() {
        return getYandexMapActivity().getFinishSearchControlPlaceholderContent();
    }

    @JavascriptInterface
    public String getEmptyBalloonContent() {
        return getYandexMapActivity().getEmptyBalloonContent();
    }

    @JavascriptInterface
    public void onDistanceChange(final int distance) {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onDistanceChange(distance);
            }
        });
    }

    @JavascriptInterface
    public void onMapCenterChange(final String text, final String title, final String subtitle,
                                  final double latitude, final double longitude) {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onMapCenterChange(text, title, subtitle, latitude, longitude);
            }
        });
    }

    @JavascriptInterface
    public void onGeolocationStart() {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onGeolocationStart();
            }
        });
    }

    @JavascriptInterface
    public void onGeolocationFinish() {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onGeolocationFinish();
            }
        });
    }

    @JavascriptInterface
    public void onEndLoading(final boolean hasError) {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onEndLoading(hasError);
            }
        });
    }

    @JavascriptInterface
    public void onErrorConstructRoute() {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onErrorConstructRoute();
            }
        });
    }
    
    @JavascriptInterface
    public void onErrorSearchPoint() {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onErrorSearchPoint();
            }
        });
    }

    @JavascriptInterface
    public void onErrorGeolocation() {
        mActivity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getYandexMapActivity().onErrorGeolocation();
            }
        });
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/YandexMapJavascriptInterface.kt
package ru.p3tr0vich.fuel

import android.app.Activity
import android.os.Build
import android.webkit.JavascriptInterface
import android.webkit.WebView

import ru.p3tr0vich.fuel.utils.UtilsLog

@Suppress("unused")
internal class YandexMapJavascriptInterface(private val activity: Activity, private val webView: WebView) {

    companion object {
        const val NAME = "YandexMapJavascriptInterface"

        private var LOG_ENABLED = false

        private const val JS = "javascript:"

        private const val ZOOM_HOUSES = 17
        private const val ZOOM_CITIES = 8
    }

    interface YandexMap {
        val mapCenterLatitude: Double

        val mapCenterLongitude: Double

        val startSearchControlPlaceholderContent: String

        val finishSearchControlPlaceholderContent: String

        val emptyBalloonContent: String

        val startSearchControlLeft: Int

        val startSearchControlTop: Int

        val finishSearchControlLeft: Int

        val finishSearchControlTop: Int

        /**
         * @param distance расстояние в метрах.
         * @param time     время в пути в секундах.
         */
        fun onRouteChange(distance: Int, time: Int)

        /**
         * @param text      полное наименования географической точки.
         * @param title     основное название (напр., название города).
         * @param subtitle  дополнительное название (напр., улица и номер дома).
         * @param latitude  широта.
         * @param longitude долгота.
         */
        fun onMapCenterChange(text: String?, title: String?, subtitle: String?,
                              latitude: Double, longitude: Double)

        fun onEndLoading(hasError: Boolean)

        fun onGeolocationStart()

        fun onGeolocationFinish()

        fun onErrorSearchPoint()

        fun onErrorConstructRoute()

        fun onErrorGeolocation()
    }

    private val yandexMapActivity: YandexMap
        get() = activity as YandexMap

    val mapCenterLatitude: Double
        @JavascriptInterface
        get() = yandexMapActivity.mapCenterLatitude

    val mapCenterLongitude: Double
        @JavascriptInterface
        get() = yandexMapActivity.mapCenterLongitude

    val startSearchControlLeft: Int
        @JavascriptInterface
        get() = yandexMapActivity.startSearchControlLeft

    val startSearchControlTop: Int
        @JavascriptInterface
        get() = yandexMapActivity.startSearchControlTop

    val finishSearchControlLeft: Int
        @JavascriptInterface
        get() = yandexMapActivity.finishSearchControlLeft

    val finishSearchControlTop: Int
        @JavascriptInterface
        get() = yandexMapActivity.finishSearchControlTop

    val startSearchControlPlaceholderContent: String
        @JavascriptInterface
        get() = yandexMapActivity.startSearchControlPlaceholderContent

    val finishSearchControlPlaceholderContent: String
        @JavascriptInterface
        get() = yandexMapActivity.finishSearchControlPlaceholderContent

    val emptyBalloonContent: String
        @JavascriptInterface
        get() = yandexMapActivity.emptyBalloonContent

    init {
        if (activity !is YandexMap)
            throw ImplementException(activity, YandexMap::class.java)
    }

    private fun runJavaScript(script: String) {
        var js = script
        js = JS + js

        if (LOG_ENABLED) {
            UtilsLog.d(NAME, "runJavaScript", "script == $js")
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            webView.evaluateJavascript(js, null)
        } else {
            webView.loadUrl(js)
        }
    }

    fun setStartLocation(latitude: Double, longitude: Double) {
        runJavaScript("setStartLocation($latitude, $longitude)")
    }

    fun setZoomIn() {
        runJavaScript("setZoomIn()")
    }

    fun setZoomOut() {
        runJavaScript("setZoomOut()")
    }

    fun setZoomToHouses() {
        runJavaScript("setZoom($ZOOM_HOUSES)")
    }

    fun setZoomToCities() {
        runJavaScript("setZoom($ZOOM_CITIES)")
    }

    @JavascriptInterface
    fun onRouteChange(distance: Int, time: Int) {
        activity.runOnUiThread { yandexMapActivity.onRouteChange(distance, time) }
    }

    @JavascriptInterface
    fun onMapCenterChange(text: String, title: String, subtitle: String,
                          latitude: Double, longitude: Double) {
        activity.runOnUiThread { yandexMapActivity.onMapCenterChange(text, title, subtitle, latitude, longitude) }
    }

    @JavascriptInterface
    fun onGeolocationStart() {
        activity.runOnUiThread { yandexMapActivity.onGeolocationStart() }
    }

    @JavascriptInterface
    fun onGeolocationFinish() {
        activity.runOnUiThread { yandexMapActivity.onGeolocationFinish() }
    }

    @JavascriptInterface
    fun onEndLoading(hasError: Boolean) {
        activity.runOnUiThread { yandexMapActivity.onEndLoading(hasError) }
    }

    @JavascriptInterface
    fun onErrorConstructRoute() {
        activity.runOnUiThread { yandexMapActivity.onErrorConstructRoute() }
    }

    @JavascriptInterface
    fun onErrorSearchPoint() {
        activity.runOnUiThread { yandexMapActivity.onErrorSearchPoint() }
    }

    @JavascriptInterface
    fun onErrorGeolocation() {
        activity.runOnUiThread { yandexMapActivity.onErrorGeolocation() }
    }
}